python/rmm/_cuda/stream.pyx

```diff
@@ -94,14 +94,17 @@ cdef class Stream:
         return self.c_is_default()
 
     def _init_from_numba_stream(self, obj):
-        self._cuda_stream = <cudaStream_t>(obj.handle.value)
+        if obj.handle.value == None:
+            self._cuda_stream = cuda_stream_default.value()
+        else:
+            self._cuda_stream = <cudaStream_t><uintptr_t>(obj.handle.value)
         self._owner = obj
 
     def _init_from_cupy_stream(self, obj):
         try:
             import cupy
             if isinstance(obj, cupy.cuda.stream.Stream):
-                self._cuda_stream = <cudaStream_t>(obj.ptr)
+                self._cuda_stream = <cudaStream_t><uintptr_t>(obj.ptr)
```

Comment:
This casting and the one above are necessary, as per this PR's description. However, I'm not really sure why they're needed, I spent half a day debugging this to figure out this handles casting correctly, I'm hoping someone like @jakirkham could clarify why this is necessary. :)

Reply:
Because otherwise obj.ptr is a Python object and you'd be casting its address to a stream pointer 😅

Reply:
whoops... Sorry for the oversight

Reply:
> Because otherwise obj.ptr is a Python object and you'd be casting its address to a stream pointer 😅

Oh, I understand it now, this converts the actual value of the pointer to `uintptr_t`, otherwise it would be equivalent to `(cudaStream_t)(&obj.ptr)`. Now it makes sense, it's very subtle and easy to overlook!

> whoops... Sorry for the oversight

No worries @shwina , it's totally understandable, as per my comment above, I would have likely done the exact same.

Reply:
Well also `uintptr_t` is something that Cython knows how to convert Python `int` objects to. It's also something Cython (and C/C++) are ok with casting to a pointer type (like `void*`). So we typically use it as intermediary type to get from a Python `int` to some C/C++ pointer type.